### PR TITLE
Redesign clickhouse query routine

### PR DESCRIPTION
**Pull request by Seitanas**
This PR moves PromHouse from loading time_series to in-memory array and running in-memory scan over milions of records on each query.
Also enables prometheus Hints for resolution in samples.

TODO:
add full support for prometheus regex queries.

@vinted/sre, @ernestas-vinted

### CODE REVIEW

File: handlers/prom.go [hunk 1]
```diff
@@ -243,23 +240,25 @@ func (p *PromAPI) read(rw http.ResponseWriter, req *http.Request) (info string,
...
 		info += fmt.Sprintf("%d: %s, ", i+1, q)
 	}
 	info = strings.TrimSuffix(info, ", ")
 	var response *prompb.ReadResponse
-	if response, err = p.storage.Read(req.Context(), queries); err != nil {
+	if response, err = p.storage.Read(req.Context(), queries, hints); err != nil {
 		return
 	}
-	p.l.Debugf("Response data:\n%s", response)
+//	p.l.Debugf("Response data:\n%s", response)
```
Comment (ernestas-vinted) [inline handlers/prom.go]: either remove or enable

File: storages/clickhouse/clickhouse.go [hunk 2]
```diff
@@ -146,63 +146,11 @@ func New(params *Params) (base.Storage, error) {
 	go func() {
 		ctx := pprof.WithLabels(context.TODO(), pprof.Labels("component", "clickhouse_reloader"))
 		pprof.SetGoroutineLabels(ctx)
-		ch.runTimeSeriesReloader(ctx)
 	}()
 
 	return ch, nil
 }
 
-func (ch *clickHouse) runTimeSeriesReloader(ctx context.Context) {
```
Comment (ernestas-vinted) [inline storages/clickhouse/clickhouse.go:155]: :+1:

File: storages/clickhouse/clickhouse.go [hunk 3]
```diff
@@ -369,7 +316,14 @@ func (ch *clickHouse) tempTableSamples(ctx context.Context, start, end int64, fi
 	return ch.scanSamples(rows)
 }
 
-func (ch *clickHouse) Read(ctx context.Context, queries []base.Query) (*prompb.ReadResponse, error) {
+func (ch *clickHouse) querySamplesNew(ctx context.Context, start, end int64, fingerprints map[uint64]struct{}) {
```
Comment (ernestas-vinted) [inline storages/clickhouse/clickhouse.go]: what does this function do ?

File: storages/clickhouse/clickhouse.go [hunk 4]
```diff
@@ -279,15 +226,15 @@ func (ch *clickHouse) scanSamples(rows *sql.Rows) ([]*prompb.TimeSeries, error)
...
 	query := fmt.Sprintf(`
-		SELECT fingerprint, timestamp_ms, value
-			FROM %s.samples
-			WHERE fingerprint IN (%s) AND timestamp_ms >= ? AND timestamp_ms <= ?
-			ORDER BY fingerprint, timestamp_ms`,
-		ch.database, placeholders[:len(placeholders)-2], // cut last ", "
+		SELECT fingerprint, timestamp_ms, value FROM (SELECT fingerprint, timestamp_ms, value FROM %s.samples WHERE fingerprint  IN (%s)
+			AND timestamp_ms >= ? AND timestamp_ms <= ? GROUP BY fingerprint, timestamp_ms, value)
+			GROUP BY fingerprint, value, timestamp_ms, intDiv(toUInt32(timestamp_ms), %v) * %v as ts ORDER BY fingerprint, timestamp_ms`,
+		ch.database, placeholders[:len(placeholders)-2], hints.StepMs, hints.StepMs, // cut last ", "
```
Comment (ernestas-vinted) [inline storages/clickhouse/clickhouse.go:237]: :+1:

File: storages/clickhouse/clickhouse.go [hunk 5]
```diff
@@ -387,36 +341,80 @@ func (ch *clickHouse) Read(ctx context.Context, queries []base.Query) (*prompb.R
...
+		label = strings.TrimSuffix(label, "}")
+		label = strings.TrimPrefix(label, "{")
+		var replacer = strings.NewReplacer(".*", "%")
+		label = replacer.Replace(label)
+		var fingerprintQuery = "SELECT DISTINCT fingerprint, labels from prometheus.time_series WHERE labels "
+		slicedLabels := strings.Split(label, ",")
+
+		for _, slice := range slicedLabels {
+			ch.l.Debugf("Slice: %s", slice)
+			if strings.Contains(slice, "=~\"") {
```
Comment (ernestas-vinted) [inline storages/clickhouse/clickhouse.go]: add few comments what/ why you are doing this for future

**Review by ernestas-vinted — APPROVED** [review]

**Comment by ernestas-vinted** [comment]
Ignore the build, upstream fails as well https://github.com/Percona-Lab/PromHouse

To test locally try:
```
make up
make test - requires clickhouse connection
make test-race
echo $? or check for Failed
```

**Comment by ernestas-vinted** [comment]
```
TODO:
add full support for prometheus regex queries.
```


Add this TODO directly in code

File: storages/clickhouse/clickhouse.go (hunk 3 above)
Comment (Seitanas) [reply storages/clickhouse/clickhouse.go]: Absolutely nothing. A left-over. Good find

File: handlers/prom.go (hunk 1 above)
Comment (Seitanas) [reply handlers/prom.go]: 👍

File: storages/clickhouse/clickhouse.go (hunk 5 above)
Comment (Seitanas) [reply storages/clickhouse/clickhouse.go]: Done.

**Comment by Seitanas** [comment]
> ```
> TODO:
> add full support for prometheus regex queries.
> ```
> 
> Add this TODO directly in code

@ernestas-vinted - done

**Review by ernestas-vinted — APPROVED** [review]

**Pull request merged**